core/pin/pin.go

```diff
@@ -49,92 +48,63 @@ func (s *Store) ProcessBlocks(ctx context.Context, c *protocol.Chain, pinName st
 				log.Error(ctx, err)
 				continue
 			}
-			err = pin.RaiseTo(ctx, block.Height)
+			err = p.raiseTo(ctx, block.Height)
 			if err != nil {
 				log.Error(ctx, err)
 			}
-		case <-ctx.Done(): // leader deposed
-			log.Error(ctx, ctx.Err())
-			return
 		}
 	}
 }
 
+func (s *Store) CreatePin(ctx context.Context, name string, height uint64) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if _, ok := s.pins[name]; ok {
+		return nil
+	}
+	const q = `
+		INSERT INTO block_processors (name, height) VALUES($1, $2)
+		ON CONFLICT(name) DO NOTHING;
+	`
+	_, err := s.db.Exec(ctx, q, name, height)
+	if err != nil {
+		return errors.Wrap(err)
+	}
+	s.pins[name] = newPin(s.db, name, height)
+	s.cond.Broadcast()
+	return nil
+}
+
 func (s *Store) LoadAll(ctx context.Context) error {
 	s.mu.Lock()
 	defer s.mu.Unlock()
-	s.pins = make(map[string]*Pin)
 	const q = `SELECT name, height FROM block_processors;`
-	err := pg.ForQueryRows(ctx, s.DB, q, func(name string, height uint64) {
-		s.pins[name] = newPin(s.DB, name, height)
+	err := pg.ForQueryRows(ctx, s.db, q, func(name string, height uint64) {
+		s.pins[name] = newPin(s.db, name, height)
 	})
+	s.cond.Broadcast()
 	return err
 }
 
-func (s *Store) WaitForAll(height uint64) <-chan struct{} {
-	ch := make(chan struct{}, 1)
+func (s *Store) pin(name string) <-chan *pin {
+	ch := make(chan *pin, 1)
 	go func() {
-		var pins []string
 		s.mu.Lock()
-		for name := range s.pins {
-			pins = append(pins, name)
-		}
-		s.mu.Unlock()
-		for _, name := range pins {
-			<-s.Pin(name).WaitForHeight(height)
+		defer s.mu.Unlock()
+		for {
+			if p, ok := s.pins[name]; ok {
+				ch <- p
+				return
+			}
+			s.cond.Wait()
```

Comment:
Maybe this is just me being too rigid, but I prefer to see a condition variable loop in a more idiomatic form:

```go
s.mu.Lock()
defer s.mu.Unlock()
for s.pins[name] == nil {
	s.cond.Wait()
}
c <- s.pins[name]
```

IMO saving the extra map lookup isn't worth the difference in readability.